### PR TITLE
bpo-39573: Finish converting to new Py_IS_TYPE macro.

### PR DESCRIPTION
**Pull request by petdance**
Updates remaining files to use Py_IS_TYPE:

```
 Include/py_curses.h                |  2 +-
 Modules/_asynciomodule.c           |  2 +-
 Modules/_collectionsmodule.c       |  2 +-
 Modules/_elementtree.c             |  2 +-
 Modules/_io/bufferedio.c           | 12 ++++++------
 Modules/_io/stringio.c             |  2 +-
 Modules/_io/textio.c               | 16 ++++++++--------
 Modules/_pickle.c                  |  4 ++--
 Modules/_threadmodule.c            |  4 ++--
 Modules/cjkcodecs/multibytecodec.h |  2 +-
 Modules/itertoolsmodule.c          |  2 +-
 Objects/abstract.c                 |  2 +-
 Objects/descrobject.c              |  4 ++--
 Objects/exceptions.c               |  4 ++--
 Objects/genobject.c                |  6 +++---
 Objects/listobject.c               | 22 +++++++++++-----------
 Objects/namespaceobject.c          |  2 +-
 Objects/tupleobject.c              |  2 +-
 Objects/typeobject.c               |  8 ++++----
 Objects/unicodeobject.c            |  4 ++--
 Python/ceval.c                     |  4 ++--
 Python/errors.c                    |  2 +-
 Python/importdl.c                  |  2 +-
```

https://bugs.python.org/issue39573

### CODE REVIEW

File: Include/py_curses.h [hunk 1]
```diff
@@ -64,7 +64,7 @@ typedef struct {
     char *encoding;
 } PyCursesWindowObject;
 
-#define PyCursesWindow_Check(v)  (Py_TYPE(v) == &PyCursesWindow_Type)
+#define PyCursesWindow_Check(v) Py_IS_TYPE((v), &PyCursesWindow_Type)
```
Comment (vstinner) [inline Include/py_curses.h]: `(v)` parenthesis looks useless:

```suggestion
#define PyCursesWindow_Check(v) Py_IS_TYPE(v, &PyCursesWindow_Type)
```

File: Modules/_io/bufferedio.c [hunk 2]
```diff
@@ -1795,8 +1795,8 @@ _io_BufferedWriter___init___impl(buffered *self, PyObject *raw,
     _bufferedwriter_reset_buf(self);
     self->pos = 0;
 
-    self->fast_closed_checks = (Py_TYPE(self) == &PyBufferedWriter_Type &&
-                                Py_TYPE(raw) == &PyFileIO_Type);
+    self->fast_closed_checks = Py_IS_TYPE(self, &PyBufferedWriter_Type) &&
+                               Py_IS_TYPE(raw, &PyFileIO_Type);
```
Comment (vstinner) [inline Modules/_io/bufferedio.c]: Please keep surrounding parenthesis:

```suggestion
    self->fast_closed_checks = (Py_IS_TYPE(self, &PyBufferedWriter_Type) &&
                                Py_IS_TYPE(raw, &PyFileIO_Type));
```

File: Modules/_io/bufferedio.c [hunk 3]
```diff
@@ -2309,8 +2309,8 @@ _io_BufferedRandom___init___impl(buffered *self, PyObject *raw,
     _bufferedwriter_reset_buf(self);
     self->pos = 0;
 
-    self->fast_closed_checks = (Py_TYPE(self) == &PyBufferedRandom_Type &&
-                                Py_TYPE(raw) == &PyFileIO_Type);
+    self->fast_closed_checks = Py_IS_TYPE(self, &PyBufferedRandom_Type) &&
+                               Py_IS_TYPE(raw, &PyFileIO_Type);
```
Comment (vstinner) [inline Modules/_io/bufferedio.c]: Ditto.
```suggestion
    self->fast_closed_checks = (Py_IS_TYPE(self, &PyBufferedRandom_Type) &&
                                Py_IS_TYPE(raw, &PyFileIO_Type));
```

File: Modules/itertoolsmodule.c [hunk 4]
```diff
@@ -527,7 +527,7 @@ teedataobject_traverse(teedataobject *tdo, visitproc visit, void * arg)
 static void
 teedataobject_safe_decref(PyObject *obj)
 {
-    while (obj && Py_TYPE(obj) == &teedataobject_type &&
+    while (obj && Py_IS_TYPE((obj), &teedataobject_type) &&
```
Comment (vstinner) [inline Modules/itertoolsmodule.c]:
```suggestion
    while (obj && Py_IS_TYPE(obj, &teedataobject_type) &&
```

File: Objects/descrobject.c [hunk 5]
```diff
@@ -1178,7 +1178,7 @@ typedef struct {
     PyObject *self;
 } wrapperobject;
 
-#define Wrapper_Check(v) (Py_TYPE(v) == &_PyMethodWrapper_Type)
+#define Wrapper_Check(v) Py_IS_TYPE((v), &_PyMethodWrapper_Type)
```
Comment (vstinner) [inline Objects/descrobject.c]:
```suggestion
#define Wrapper_Check(v) Py_IS_TYPE(v, &_PyMethodWrapper_Type)
```

File: Objects/genobject.c [hunk 6]
```diff
@@ -1216,10 +1216,10 @@ static PyAsyncGenASend *ag_asend_freelist[_PyAsyncGen_MAXFREELIST];
 static int ag_asend_freelist_free = 0;
 
 #define _PyAsyncGenWrappedValue_CheckExact(o) \
-                    (Py_TYPE(o) == &_PyAsyncGenWrappedValue_Type)
+                    Py_IS_TYPE((o), &_PyAsyncGenWrappedValue_Type)
```
Comment (vstinner) [inline Objects/genobject.c]:
```suggestion
                    Py_IS_TYPE(o, &_PyAsyncGenWrappedValue_Type)
```

File: Objects/genobject.c [hunk 7]
```diff
@@ -1216,10 +1216,10 @@ static PyAsyncGenASend *ag_asend_freelist[_PyAsyncGen_MAXFREELIST];
...
 
 #define PyAsyncGenASend_CheckExact(o) \
-                    (Py_TYPE(o) == &_PyAsyncGenASend_Type)
+                    Py_IS_TYPE((o), &_PyAsyncGenASend_Type)
```
Comment (vstinner) [inline Objects/genobject.c]:
```suggestion
                    Py_IS_TYPE(o, &_PyAsyncGenASend_Type)
```

File: Modules/_io/bufferedio.c [hunk 8]
```diff
@@ -1449,8 +1449,8 @@ _io_BufferedReader___init___impl(buffered *self, PyObject *raw,
         return -1;
     _bufferedreader_reset_buf(self);
 
-    self->fast_closed_checks = (Py_TYPE(self) == &PyBufferedReader_Type &&
-                                Py_TYPE(raw) == &PyFileIO_Type);
+    self->fast_closed_checks = Py_IS_TYPE(self, &PyBufferedReader_Type) &&
+                               Py_IS_TYPE(raw, &PyFileIO_Type);
```
Comment (vstinner) [inline Modules/_io/bufferedio.c]:
```suggestion
    self->fast_closed_checks = (Py_IS_TYPE(self, &PyBufferedReader_Type) &&
                                Py_IS_TYPE(raw, &PyFileIO_Type));
```

**Review by vstinner — COMMENTED** [review]
First round of review: the parenthesis review :-)

File: Modules/_io/bufferedio.c (hunk 8 above)
Comment (petdance) [reply Modules/_io/bufferedio.c]: Should I make a new PR with these changes?

**Comment by vstinner** [comment]
@petdance: Thanks, merged. There are code like `(Py_TYPE(other) != &keyobject_type)`. Do you want to convert them to `!Py_IS_TYPE(obj, type)` as well?

**Comment by petdance** [comment]
>  Do you want to convert them to !Py_IS_TYPE(obj, type) as well?

Sure, I'll take care of them.

**Comment by petdance** [comment]
@vstinner I've updated the negated checks to use `Py_IS_TYPE` as well over at #18789.

**Pull request merged**